### PR TITLE
Decorators for validation

### PR DESCRIPTION
**Pull request by edomora97**
To reduce code duplication I've implemented a simple to use validation process for the requests.

Before calling an handler a series of validation methods is called. These methods will check many things, for example if the contest is running, if the provided id is valid, and many other things including logging the user ip.

To enable this process you only need to add a decorator to the endpoint, like
```python
@Validators.validate_during_contest
@Validators.register_user_ip
@Validators.validate_token
@Validators.validate_task
def generate_input(self, task, user):
    # task and user are the rows from the db corresponding to the request parameters
    # task and token
```

A list of the implemented decorators can be found in the `src/validators.py` file.

### CODE REVIEW

File: src/handlers/admin_handler.py [hunk 1]
```diff
@@ -88,27 +73,27 @@ def start(self, admin_token:str, _ip):
...
+    @Validators.validate_admin_only
+    def set_extra_time(self, extra_time:int, token:str=None):
         """
         POST /admin/set_extra_time
         """
-        self._validate_token(admin_token, _ip)
-
         if token is None:
             Database.set_meta("extra_time", extra_time)
+        elif Database.get_user(token) is None:
```
Comment (veluca93) [inline src/handlers/admin_handler.py]: Should we use a validator for this too? (by passing a boolean parameter "can_be_none" to the validator, for example)

File: src/validators.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+from src.config import Config
+from src.database import Database
+from src.logger import Logger
+from .handlers.base_handler import BaseHandler
+
+
+class Validators:
+
+    @staticmethod
+    def validate_during_contest(handler):
```
Comment (veluca93) [inline src/validators.py]: Shouldn't we call this during_contest?

File: src/validators.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        def handle(*args, **kwargs):
+            token = Validators._guess_token(**kwargs)
+            Validators._ensure_contest_running(token)
+            return handler(*args, **kwargs)
+        return BaseHandler.initialize_request_params(handle, handler)
+
+    @staticmethod
+    def validate_admin_only(handler):
```
Comment (veluca93) [inline src/validators.py]: ... and this admin_only?

File: test/handlers/test_base_handler.py [hunk 4]
```diff
@@ -212,6 +217,21 @@ def test_call_general_attrs(self, name_mock, content_mock, ip_mock):
...
+
+        res = handler._call(handler.file, {}, request)
+        self.assertEqual("foo", res)
+
+
+
+
+
+
+
```
Comment (veluca93) [inline test/handlers/test_base_handler.py]: Why so many blank lines?

File: src/handlers/admin_handler.py (hunk 1 above)
Comment (edomora97) [reply src/handlers/admin_handler.py]: Good idea

File: test/handlers/test_base_handler.py (hunk 4 above)
Comment (edomora97) [reply test/handlers/test_base_handler.py]: Ops :joy:

Comment (edomora97) [reply test/handlers/test_base_handler.py]: Because I'm an idiot ahah

**Comment by edomora97** [comment]
[Presto (quando il branch decorators verrà mergiato) ci saranno delle API rotte](https://trello.com/c/cLhkFC9W/33-presto-quando-il-branch-decorators-verra-mergiato-ci-saranno-delle-api-rotte)

**Pull request merged**